optimum/neuron/trainers.py

```diff
@@ -31,12 +50,227 @@
+            self.tmp_neuron_cache_state += diff
+            if self.use_neuron_cache:
+                for path in diff:
+                    path_in_cache = self.full_path_to_path_in_cache(path)
+                    path_in_original_cache = self.neuron_cache_path / path_in_cache
+                    path_in_original_cache.parent.mkdir(parents=True, exist_ok=True)
+                    if path_in_original_cache.exists():
+                        continue
+                    shutil.copy(path, path_in_original_cache)
+
+        return found_in_cache
+
+    def synchronize_temporary_neuron_cache_state(self) -> List[Path]:
+        current_files_in_neuron_cache = list_files_in_neuron_cache(
+            self.tmp_neuron_cache_path, only_relevant_files=True
+        )
+        diff = [p for p in current_files_in_neuron_cache if p not in self.tmp_neuron_cache_state]
+        self.tmp_neuron_cache_state = current_files_in_neuron_cache
+        return diff
+
+    def synchronize_temporary_neuron_cache(self):
+        for neuron_hash, files in self.neuron_hash_to_files.items():
+
+            def local_path_to_path_in_repo(path):
+                return path_after_folder(path, f"USER_neuroncc-{neuron_hash.neuron_compiler_version}")
+
+            for path in files:
+                push_to_cache_on_hub(neuron_hash, path, local_path_to_path_in_repo=local_path_to_path_in_repo)
+                if self.use_neuron_cache:
+                    path_in_cache = self.full_path_to_path_in_cache(path)
+                    target_file = self.neuron_cache_path / path_in_cache
+                    target_file.parent.mkdir(parents=True, exist_ok=True)
+                    shutil.copy(path, self.neuron_cache_path / path_in_cache)
+
+        if self.use_neuron_cache:
+            self._update_cache_stats(self.neuron_cache_path)
+
+        for neuron_hash in self.neuron_hash_to_files:
+            self.neuron_hash_to_files[neuron_hash] = []
+
+    def on_step_end(self, args: "TrainingArguments", state: "TrainerState", control: "TrainerControl", **kwargs):
+        """
+        Event called at the end of a training step. If using gradient accumulation, one training step might take
+        several inputs.
+        """
+        model = kwargs["model"]
+        state = self.prepare_state(state)
+        neuron_hash = self.neuron_hash_for_model(args, model, state.last_inputs, try_to_fetch_cached_model=True)
+        diff = self.synchronize_temporary_neuron_cache_state()
+        self.neuron_hash_to_files[neuron_hash].extend(diff)
+
+    def on_save(self, args: "TrainingArguments", state: TrainerState, control: "TrainerControl", **kwargs):
+        """
+        Event called after a checkpoint save.
+        """
+        self.synchronize_temporary_neuron_cache()
+
+    def on_step_middle(self, args: "TrainingArguments", state: TrainerState, control: "TrainerControl", **kwargs):
+        model = kwargs["model"]
```

Comment:
I created the "middle" step because we needed an event right before running inference, where the inputs are accessible, but nothing has been computed yet. This way, we can check if compilation will need to be done, before actually doing it, and thus downloading the proper cache files beforehand.